vault/logical_system.go

```diff
@@ -5542,6 +5545,397 @@ func checkListingVisibility(visibility ListingVisibilityType) error {
+		return logical.ErrorResponse("missing id parameter value"), nil
+	}
+
+	id = idValue.(string)
+
+	if err = b.Core.uiConfig.DeleteCustomMessage(ctx, id); err != nil {
+		return logical.ErrorResponse("failed to delete custom message: %s", err), nil
+	}
+
+	return nil, nil
+}
+
+func (b *SystemBackend) handleUpdateCustomMessage(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var err error
+
+	var messageId string
+	messageIdValue, ok, err := d.GetOkErr("id")
+	if err != nil {
+		return logical.ErrorResponse("invalid id parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing id parameter value"), nil
+	}
+
+	messageId = messageIdValue.(string)
+
+	var title string
+	titleValue, ok, err := d.GetOkErr("title")
+	if err != nil {
+		return logical.ErrorResponse("invalid title parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing title parameter value"), nil
+	}
+
+	title = titleValue.(string)
+
+	var authenticated bool
+	authenticatedValue, ok, err := d.GetOkErr("authenticated")
+	if err != nil {
+		return logical.ErrorResponse("invalid authenticated parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing authenticated parameter value"), nil
+	}
+
+	authenticated = authenticatedValue.(bool)
+
+	var messageType string
+	messageTypeValue, ok, err := d.GetOkErr("type")
+	if err != nil {
+		return logical.ErrorResponse("invalid type parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing type parameter value"), nil
+	}
+
+	messageType = messageTypeValue.(string)
+
+	var message string
+	messageValue, ok, err := d.GetOkErr("message")
+	if err != nil {
+		return logical.ErrorResponse("invalid message parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing message parameter value"), nil
+	}
+
+	message = messageValue.(string)
+
+	var link map[string]any
+	linkValue, ok, err := d.GetOkErr("link")
+	if err != nil {
+		return logical.ErrorResponse("invalid link parameter value"), nil
+	}
+
+	if ok {
+		link = linkValue.(map[string]any)
+	}
+
+	var options map[string]any
+	optionsValue, ok, err := d.GetOkErr("options")
+	if err != nil {
+		return logical.ErrorResponse("invalid options parameter value"), nil
+	}
+
+	if ok {
+		options = optionsValue.(map[string]any)
+	}
+
+	var startTime time.Time
+	startTimeValue, ok, err := d.GetOkErr("start_time")
+	if err != nil {
+		return logical.ErrorResponse("invalid start_time parameter value"), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing start_time parameter value"), nil
+	}
+
+	startTime = startTimeValue.(time.Time)
+
+	var endTime time.Time
+	endTimeValue, ok, err := d.GetOkErr("end_time")
+	if err != nil {
+		return logical.ErrorResponse("invalid end_time parameter value"), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing end_time parameter value"), nil
+	}
+
```

Comment:
Likewise the time validation here (perhaps it makes sense to make a helper method)